AK/URLParser.cpp

```diff
@@ -44,10 +46,184 @@ static Optional<DeprecatedString> parse_opaque_host(StringView input)
+    u32 output;
+    if (radix == 8)
+        output = StringUtils::convert_to_uint_from_octal(input).release_value();
+    else if (radix == 10)
+        output = input.to_uint().release_value();
+    else if (radix == 16)
+        output = StringUtils::convert_to_uint_from_hex(input).release_value();
+    else
+        VERIFY_NOT_REACHED();
+
+    // 9. Return (output, validationError).
+    return ParsedIPV4Number { output, validation_error };
+}
+
+// https://url.spec.whatwg.org/#concept-ipv4-parser
+static Optional<u32> parse_ipv4_address(StringView input)
+{
+    // 1. Let parts be the result of strictly splitting input on U+002E (.).
+    auto parts = input.split_view("."sv, SplitBehavior::KeepEmpty);
+
+    // 2. If the last item in parts is the empty string, then:
+    if (parts.last().is_empty()) {
+        // 1. IPv4-empty-part validation error.
+        report_validation_error();
+
+        // 2. If parts’s size is greater than 1, then remove the last item from parts.
+        if (parts.size() > 1)
+            parts.take_last();
+    }
+
+    // 3. If parts’s size is greater than 4, IPv4-too-many-parts validation error, return failure.
+    if (parts.size() > 4) {
+        report_validation_error();
+        return {};
+    }
+
+    // 4. Let numbers be an empty list.
+    Vector<u32, 4> numbers;
+
+    // 5. For each part of parts:
+    for (auto const& part : parts) {
+        // 1. Let result be the result of parsing part.
+        auto const result = parse_ipv4_number(part);
+
+        // 2. If result is failure, IPv4-non-numeric-part validation error, return failure.
+        if (!result.has_value()) {
+            report_validation_error();
+            return {};
+        }
+
+        // 3. If result[1] is true, IPv4-non-decimal-part validation error.
+        if (result->validation_error)
+            report_validation_error();
+
+        // 4. Append result[0] to numbers.
+        numbers.append(result->number);
+    }
+
+    // 6. If any item in numbers is greater than 255, IPv4-out-of-range-part validation error.
+    // 7. If any but the last item in numbers is greater than 255, then return failure.
+    for (size_t i = 0; i < numbers.size(); ++i) {
+        if (numbers[i] > 255) {
+            report_validation_error();
+            if (i != numbers.size() - 1)
+                return {};
+        }
+    }
+
+    // 8. If the last item in numbers is greater than or equal to 256(5 − numbers’s size), then return failure.
+    if (numbers.last() >= pow<size_t>(256, 5 - numbers.size()))
+        return {};
+
+    // 9. Let ipv4 be the last item in numbers.
+    auto ipv4 = numbers.last();
+
+    // 10. Remove the last item from numbers.
+    numbers.take_last();
+
+    // 11. Let counter be 0.
+    u8 counter = 0;
+
+    // 12. For each n of numbers:
+    for (u32 n : numbers) {
+        // 1. Increment ipv4 by n × 256(3 − counter).
+        ipv4 += n * pow<size_t>(256, 3 - counter);
```

Comment:
Probably isn't going to show up in profiles, but the `pow<size_t>` invocations in this file *could* be replaced with bitwise math
```c++
ipv4 |= n << (8 * (3 - counter))
```
(not asking you to replace this here, just making a note if this does become hot)